File: number-game/number-game.cpp
#include<iostream>
#include<cstring>
#include<vector>
using namespace std;

const int MAX = 9999999;
int n, board[50], result[50];
int cache[50][50];
//-1000 -1000 -3 -1000 -1000 -1000
//100 -1000 -1000 100 -1000 -1000 1100
//7 -5 8 5 1 -4 -8 6 7 9

int NUMBERGAME(int left, int right) {
	if(left > right) 
            return 0;
	int& ret = cache[left][right];
	if(ret != MAX) 
           return ret;

	ret = max(board[left] - NUMBERGAME(left + 1, right), board[right] - NUMBERGAME(left, right - 1));
  //  cout<<"ret = "<< ret <<" left = "<< left <<" right = "<< right << endl;
	if(right - left >= 2) 
    {
		ret = max(ret, -NUMBERGAME(left + 2, right));
		ret = max(ret, -NUMBERGAME(left, right - 2));
	}
	return ret;
}

int main() {
    int cases;
    cin >> cases;
    for(int i = 0; i < cases; ++i) 
    {
    	cin >> n;
    	for(int j = 0; j < n; j++) 
                cin >> board[j];
    	for(int j = 0; j < n; j++)
    		for(int k = 0; k < n; k++)
    			cache[j][k] = MAX;
    	
        result[i] = NUMBERGAME(0, n-1);
    }
    
    for(int i = 0; i < cases; ++i) 
    {
            cout << result[i] << endl;
    }
//    system("PAUSE");
}

File: number-game/number-game.py
import sys

readl = lambda : sys.stdin.readline().strip()

EMPTY = -9999999
table = []
sequence = []

def max(num1, num2):
    if num1 > num2 : return num1
    else : return num2

def game(left, right):
    global table, sequence

    if left > right : return 0
    
    if table[left][right] != EMPTY : return table[left][right]

    lsum = sequence[left] - game(left+1, right)
    rsum = sequence[right] - game(left, right-1)
    
    result = max(lsum, rsum)
    if (right - left + 1) >=2 :
        result = max(result, -game(left+2, right))
        result = max(result, -game(left, right-2))
    
    table[left][right] = result
    return result


testCases = int(readl())

for _ in range(testCases) :
    numbers = int(readl())
    sequence = map(int, readl().split())
    
    table = []
    for _ in range(numbers) :
        table.append([EMPTY] * numbers)
    
    result = game(0, numbers-1)
    print(result)
